crategen/validators.py

```diff
@@ -0,0 +1,109 @@
+from pydantic import ValidationError
+from typing import Union
+from .models import WRROCProcess, WRROCWorkflow, WRROCProvenance
+from urllib.parse import urlparse
+
+def validate_wrroc(data: dict) -> Union[WRROCProvenance, WRROCWorkflow, WRROCProcess]:
+    """
+    Validate that the input data is a valid WRROC entity and determine which profile it adheres to.
+    
+    This function attempts to validate the input data against the WRROCProvenance model first.
+    If that validation fails, it attempts validation against the WRROCWorkflow model.
+    If that also fails, it finally attempts validation against the WRROCProcess model.
+    
+    Args:
+        data (dict): The input data to validate.
+    
+    Returns:
+        Union[WRROCProvenance, WRROCWorkflow, WRROCProcess]: The validated WRROC data, indicating the highest profile the data adheres to.
+    
+    Raises:
+        ValueError: If the data does not adhere to any of the WRROC profiles.
+    """
+    # Convert '@id' to 'id' for validation purposes
+    if '@id' in data:
+        data['id'] = data.pop('@id')
+
+    errors = []
+
+    try:
+        return WRROCProvenance(**data)
+    except ValidationError as e:
+        errors.extend(e.errors())
+
+    try:
+        return WRROCWorkflow(**data)
+    except ValidationError as e:
+        errors.extend(e.errors())
+
+    try:
+        return WRROCProcess(**data)
+    except ValidationError as e:
+        errors.extend(e.errors())
+        raise ValueError(f"Invalid WRROC data: {errors}")
+
+def validate_wrroc_tes(data: dict) -> WRROCProcess:
```

Comment:
Add type hint for `data`.

Reply:
I've implemented the validators directly in the `convert_from_wrroc` methods of both TES and WES converters. This ensures that all data is properly validated before the conversion process.